include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+{
+	return irr_glsl_blinn_phong_cos_eval_DG(uf.NdotH, i.NdotV, s.NdotL, n, a2);
+}
+
+float irr_glsl_blinn_phong_cos_eval_DG(in float NdotH, in float TdotH, in float BdotH, in float NdotL, in float TdotL, in float BdotL, in float NdotV, in float TdotV, in float BdotV, in float nx, in float ny, in float ax2, in float ay2)
+{
+	return irr_glsl_blinn_phong_cos_eval_DG_wo_clamps(NdotH, NdotH * NdotH, TdotH * TdotH, BdotH * BdotH, TdotL * TdotL, BdotL * BdotL, TdotV * TdotV, BdotV * BdotV, NdotV * NdotV, NdotL * NdotL, nx, ny, ax2, ay2);
+}
+float irr_glsl_blinn_phong_cos_eval_DG(in irr_glsl_LightSample s, in irr_glsl_AnisotropicMicrofacetCache uf, in irr_glsl_AnisotropicViewSurfaceInteraction i, in float nx, in float ny, in float ax2, in float ay2)
+{
+	return irr_glsl_blinn_phong_cos_eval_DG(
+		uf.isotropic.NdotH,
+		uf.TdotH,
+		uf.BdotH,
+		s.NdotL,
+		s.TdotL,
+		s.BdotL,
+		i.isotropic.NdotV,
+		i.TdotV,
+		i.BdotV,
+		nx, ny, ax2, ay2
+	);
+}
+
+#ifdef GEN_CHOICE_STREAM
+void instr_eval_and_pdf_execute(in instr_t instr, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache _uf, inout bool ts_flag)
+{
+	uint op = instr_getOpcode(instr);
+
+	//speculative execution
+	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+	params_t params = instr_getParameters(instr, bsdf_data);
+	uint ndf = instr_getNDF(instr);
+	float a = params_getAlpha(params);
+	float a2 = a*a;
+	float n = irr_glsl_alpha2_to_phong_exp(a2);
+#ifndef ALL_ISOTROPIC_BXDFS
+	float ay = params_getAlphaV(params);
+	float ay2 = ay*ay;
+	float ny = irr_glsl_alpha2_to_phong_exp(ay2);
+#endif
+	float bxdf_eval_scalar_part = 0.0;
+	float pdf = 0.0;
+
+	const mat2x3 ior = bsdf_data_decodeIoR(bsdf_data,op);
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons)
+
+#ifndef NO_TWOSIDED
+	handleTwosided(ts_flag, instr, s, _uf);
+#endif
+
+	float cosFactor = is_bsdf ? abs(s.NdotL):max(s.NdotL,0.0);
+
+	irr_glsl_AnisotropicMicrofacetCache uf;
+	//here actually using stronger check for BSDF because it's probably worth it
+	if (op_isBSDF(op) && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+	{
+		float orientedEta, rcpOrientedEta;
+		irr_glsl_getOrientedEtas(orientedEta, rcpOrientedEta, currInteraction.isotropic.NdotV, ior[0].x);
+		const bool valid = irr_glsl_calcAnisotropicMicrofacetCache(uf, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, orientedEta, rcpOrientedEta);
+		//assert(valid);
+	}
+	else
+	{
+		uf = _uf;
+	}
+
+	if (cosFactor>FLT_MIN && op_hasSpecular(op))//does cos>0 check even has any point here? L comes from a sample..
+	{
+#ifdef OP_DIFFUSE
+		if (op==OP_DIFFUSE) {
+			pdf = irr_glsl_oren_nayar_pdf_wo_clamps(cosFactor);
+		} else
+#endif
+
+#ifdef OP_DIFFTRANS
+		if (op==OP_DIFFTRANS) {
+			pdf = irr_glsl_lambertian_transmitter_pdf_wo_clamps(cosFactor);
```

Comment:
you can unify these, use `wo_clamps_pi_factored_out` and then `/= is_bsdf ? (2.0*PI):PI;`